### PR TITLE
V3.1.0

### PR DESCRIPTION
**Pull request by SC426**
[Upstream](https://github.com/networkx/networkx/tree/networkx-3.1)
[Changelog](https://networkx.org/documentation/latest/release/release_dev.html)

## Actions
- Updated version to 3.1 (3.1.0 causes a URL error)
- Updated run script to include `--no-build-isoloation`
- Remove outdated tests
- Add required and extra packages for testing

## Questions
- Should we still leave out these required/optional packages Per the comment left in the `meta.yaml` by @skupr-anaconda?
- If so, is my formatting correct?
- I will clean up the `meta.yaml` once these questions are answered.

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -30,6 +32,18 @@ requirements:
     #- pandas >=1.3
     # Optional lxml, pygraphviz, pydot, gdal packages provide additional functionality
 
+    # Steve:
+    # Do we want to add the following packages to this recipe?
```
Comment (rafaelmartins) [inline recipe/meta.yaml]: it is usually not a good idea to enforce optional dependencies, just add any optional dependencies together with `networkx` when packaging anything that needs them.

Comment (SC426) [reply recipe/meta.yaml]: Ok I've removed all the extras and pushed the changes.

File: recipe/meta.yaml [hunk 2]
```diff
@@ -1,17 +1,17 @@
...
 package:
   name: {{ name|lower }}
   version: {{ version }}
 
 source:
   url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
-  sha256: 5e53f027c0d567cf1f884dbb283224df525644e43afd1145d64c9d88a3584762
+  sha256: de346335408f84de0eada6ff9fafafff9bcda11f0a0dfaa931133debb146ab61
 build:
   number: 1
```
Comment (cbouss) [inline recipe/meta.yaml]:
```suggestion
  number: 0
```

File: recipe/meta.yaml [hunk 3]
```diff
@@ -1,17 +1,17 @@
...
   skip: True  # [py<38]
-  script: {{ PYTHON }} -m pip install . -vvv --no-deps
+  script: {{ PYTHON }} -m pip install . -vvv --no-deps --no-build-isolation
 
 requirements:
   host:
```
Comment (cbouss) [inline recipe/meta.yaml:17]: You could add a run_constrained section with the content of:
https://github.com/networkx/networkx/blob/networkx-3.1/requirements/default.txt and https://github.com/networkx/networkx/blob/networkx-3.1/requirements/extra.txt

Comment (cbouss) [reply recipe/meta.yaml:17]: And remove the big comment form the run section.

Comment (SC426) [reply recipe/meta.yaml:17]: Ok I've made those changes

**Review by cbouss — APPROVED** [review]
LGTM!

**Comment by skupr-anaconda** [comment]
Overall looks good but could you test the downstream packages https://metayaml-conda.fly.dev/metayaml/reverse_dependencies?name=networkx if they work correctly with networkx 3.1.0?
You can use the stage channel or add `downstreams` to the `test` section and build it locally to catch errors:
```
test:
  downstreams:
    - mapclassify 2.5.0
    - odo 0.5.1
    - orange3 3.32.0
    - pomegranate 0.14.4
    - python-louvain 0.15
    - scikit-image 0.20.0
    - visions 0.7.5
```

File: recipe/meta.yaml (hunk 3 above)
Comment (boldorider4) [reply recipe/meta.yaml:17]: Why do we need to move all requirements to the `run_constrained` section? Perhaps my understanding of `run_constrained` is limited.

Comment (skupr-anaconda) [reply recipe/meta.yaml:17]: If the user will install a package mentioned in `run_constrained` to the  environment then the package will have a the range of the pinnings from `run_constrained`

Comment (boldorider4) [reply recipe/meta.yaml:17]: but why are they all in `run_constrained` and not at least the default requirements in the `run` section?

Comment (rafaelmartins) [reply recipe/meta.yaml:17]: this way they are optional dependencies, and the constraint will still ensure that they are in the valid range if the user install the optional dependency manually

Comment (boldorider4) [reply recipe/meta.yaml:17]: I understand now what `run_constrained` does, but I'm still not convinced by the fact that packages that are listed as default requirements are considered optional dependencies.

**Comment by SC426** [comment]
@skupr-anaconda it looks like tests for `scikit-image` failed:
`18 failed, 7934 passed, 33 skipped, 136 deselected, 341 warnings in 189.17s (0:03:09) =
TESTS FAILED: scikit-image-0.20.0-py311h313beb8_0.tar.bz2`

File: recipe/meta.yaml (hunk 3 above)
Comment (rafaelmartins) [reply recipe/meta.yaml:17]: the thing is that every dependency here is optional, even the ones in default.txt. you can see that install_requires is empty in setup.py: https://github.com/networkx/networkx/blob/networkx-3.1/setup.py#L167

To install the deps from default.txt with pip you'd use `pip install networkx[default]`

Comment (cbouss) [reply recipe/meta.yaml:17]: The source of truth of what is required is here: https://github.com/networkx/networkx/blob/networkx-3.1/setup.py#L194 and https://github.com/networkx/networkx/blob/networkx-3.1/setup.py#L167
Optional are right below: https://github.com/networkx/networkx/blob/networkx-3.1/setup.py#L168

**Comment by cbouss** [comment]
> @skupr-anaconda it looks like tests for `scikit-image` failed: `18 failed, 7934 passed, 33 skipped, 136 deselected, 341 warnings in 189.17s (0:03:09) = TESTS FAILED: scikit-image-0.20.0-py311h313beb8_0.tar.bz2`

The failures may not be due to networkx update.
What if you run scikit-image tests against the current version of networkx in default?

**Comment by SC426** [comment]
@cbouss Here are those results:
`=========================== short test summary info ============================
FAILED data/tests/test_data.py::test_eagle - AssertionError:
FAILED io/tests/test_collection.py::TestImageCollection::test_custom_load_func_sequence
FAILED registration/tests/test_phase_cross_correlation.py::test_disambiguate_2d[100-100]
FAILED registration/tests/test_phase_cross_correlation.py::test_disambiguate_2d[100--100]
FAILED registration/tests/test_phase_cross_correlation.py::test_disambiguate_2d[100-350]
FAILED registration/tests/test_phase_cross_correlation.py::test_disambiguate_2d[100--350]
FAILED registration/tests/test_phase_cross_correlation.py::test_disambiguate_2d[-100-100]
FAILED registration/tests/test_phase_cross_correlation.py::test_disambiguate_2d[-100--100]
FAILED registration/tests/test_phase_cross_correlation.py::test_disambiguate_2d[-100-350]
FAILED registration/tests/test_phase_cross_correlation.py::test_disambiguate_2d[-100--350]
FAILED registration/tests/test_phase_cross_correlation.py::test_disambiguate_2d[350-100]
FAILED registration/tests/test_phase_cross_correlation.py::test_disambiguate_2d[350--100]
FAILED registration/tests/test_phase_cross_correlation.py::test_disambiguate_2d[350-350]
FAILED registration/tests/test_phase_cross_correlation.py::test_disambiguate_2d[350--350]
FAILED registration/tests/test_phase_cross_correlation.py::test_disambiguate_2d[-350-100]
FAILED registration/tests/test_phase_cross_correlation.py::test_disambiguate_2d[-350--100]
FAILED registration/tests/test_phase_cross_correlation.py::test_disambiguate_2d[-350-350]
FAILED registration/tests/test_phase_cross_correlation.py::test_disambiguate_2d[-350--350]
= 18 failed, 7934 passed, 33 skipped, 136 deselected, 341 warnings in 124.74s (0:02:04) =`

File: recipe/meta.yaml (hunk 3 above)
Comment (skupr-anaconda) [reply recipe/meta.yaml:17]: @boldorider4 If you will add those dependencies to `run` then downstream packages will have a dependency hell. It has already happened with **v2.6.2** https://github.com/AnacondaRecipes/networkx-feedstock/commit/e0063b65a3d992664d96c3c0d275ef8052b2dee6 and the maintainers decided to remove all of them from runtime

**Comment by SC426** [comment]
@skupr-anaconda After both tests it doesn't seem like networkx is causing any issues (although there are 18 unexplained failures for scikit-image on both accounts).

**Comment by skupr-anaconda** [comment]
> test_disambiguate_2d

This test is related to `numpy` https://github.com/scikit-image/scikit-image/blob/v0.20.0/skimage/registration/tests/test_phase_cross_correlation.py#L237

**Comment by SC426** [comment]
@skupr-anaconda After both tests it doesn't seem like networkx is causing any issues (although there are 18 unexplained failures for scikit-image on both accounts).

**Comment by SC426** [comment]
@skupr-anaconda Thank you for finding that test. What do you think is the best course of action now?

File: recipe/meta.yaml (hunk 3 above)
Comment (boldorider4) [reply recipe/meta.yaml:17]: got it. Thanks everyone for clarifying this.

**Review by boldorider4 — APPROVED** [review]
LGTM

**Comment by skupr-anaconda** [comment]
> @skupr-anaconda Thank you for finding that test. What do you think is the best course of action now?

@SC426 If no other issues you can go ahead with merging the PR

**Review by skupr-anaconda — APPROVED** [review]

**Pull request merged**